### PR TITLE
Woo: Update product fields

### LINKED ISSUE #1435: Woo Poducts: add support to update empty product description
PR #1432 added support to update product description. But there is no way for the user to update a blank product description at the moment. This is supported by `wp-admin` and needs to be added to the app

### PR DESCRIPTION
**Pull request by anitaa1990**
Fixes #1435. This PR adds logic to update the following product fields to the API.

#### Changes
- Adds two columns/fields `dateOnSaleFrom` and `dateOnSaleTo` to `WCProductModel` to allow for scheduling product sales.
- Adds logic to update product only if updated product fields do not match the stored product fields. This is to allow the option for users to update empty product descriptions (should they choose to do so) and also, to ensure that we do not update product fields that do not contain any changes. 
- Testing the API, I found that we cannot update an existing SKU value from another product to the product we are trying to update. So added a new error type in `ProductErrorType` if user tries to update the SKU of the product by entering an existing value.
- Added enum classes for displaying product tax status options (`CoreProductTaxStatus`), product back order options (`CoreProductBackOrders`) and product stock status options (`CoreProductStockStatus`).
- Added a new page in the example app to test updating multiple product fields.

#### Testing
- Run tests in `MockedStack_WCProductsTest`, `WCProductStoreTest`.
- Since there is a migration script added to the PR, it would be good to verify if app can be updated to this PR without any issues.
- Testing in the example app:
  - Open the app, click on `Woo` (after login, of course) -> `Products` and select a site.
  - Click on `Update Product` button.
  - Enter a product ID.
  - Modify fields displayed in the UI and verify that only the fields updated in the UI are updated to the product.
  - Verify that an error is displayed in the example app when trying to update an existing SKU (from another product).

#### Screenshots
 . 

#### Notes 
- There is no option to test updating `tax_class` and `shipping_class` fields to a product. This is because we need to add a separate API endpoints to FluxC before testing this. Added issue [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/1442) and here to address this separately.

### CODE REVIEW

**Comment by nbradbury** [comment]
This doesn't need to be done in this PR, but it would be nice if this screen relied on labels above the TextViews rather than hints. It's not clear what each fields is when filled in. For example, I don't know what the "2", "3", and "4" fields are here.

Screenshot_1575899170

**Comment by nbradbury** [comment]
If you rotate the device while the list dialog is showing, it loses its list items.

Screenshot_1575899339

**Comment by nbradbury** [comment]
If I rotate the device while the date picker is showing, the date picker disappears and I'm left with a screen where all the fields are disabled.

Screenshot_1575899532

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductRestClient.kt [hunk 1]
```diff
@@ -420,19 +425,93 @@ class ProductRestClient(
...
+            if (storedWCProductModel.stockStatus != updatedProductModel.stockStatus) {
+                body["stock_status"] = updatedProductModel.stockStatus
+            }
+            if (storedWCProductModel.stockQuantity != updatedProductModel.stockQuantity) {
+                body["stock_quantity"] = updatedProductModel.stockQuantity
+            }
+            if (storedWCProductModel.backorders != updatedProductModel.backorders) {
+                body["backorders"] = updatedProductModel.backorders
+            }
+            if (storedWCProductModel.backorders != updatedProductModel.backorders) {
```
Comment (nbradbury) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductRestClient.kt]: Backorders is duplicated here.

**Comment by nbradbury** [comment]
@anitaa1990 Code looks good - just a few minor issues to fix and I'm good to merge.

**Comment by anitaa1990** [comment]
Thanks for the quick review @nbradbury!

> It would be nice if this screen relied on labels above the TextViews rather than hints.  

Good point! Fixed in 66cd2b0

> If you rotate the device while the list dialog is showing, it loses its list items.

Fixed in 762ea3a

> If I rotate the device while the date picker is showing, the date picker disappears and I'm left with a screen where all the fields are disabled.

Fixed in 1cc77ea

Ready for another round 👍

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductRestClient.kt (hunk 1 above)
Comment (anitaa1990) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductRestClient.kt]: Fixed in 6bbdd44

**Pull request merged**